flocker/node/_deploy.py

```diff
@@ -424,6 +453,34 @@ def calculate_necessary_state_changes(self, desired_state,
         :return: A ``Deferred`` which fires with a ``IStateChange``
             provider.
         """
+        # Add missing dataset IDs to the desired configuration. Should be
+        # done elsewhere: https://clusterhq.atlassian.net/browse/FLOC-1199
+
+        # We don't want to mutate the desired configuration, so do a
+        # deepcopy (PMap doesn't support copy.deepcopy):
+        desired_state = loads(dumps(desired_state))
+
+        datasets_with_no_id = []
+        for node in desired_state.nodes:
+            for application in node.applications:
+                if application.volume:
+                    dataset = application.volume.dataset
+                    if dataset.dataset_id is None:
+                        datasets_with_no_id.append(dataset)
+        current_datasets_by_name = {}
+        for node in current_cluster_state.nodes:
+            for application in node.applications:
+                if application.volume:
+                    dataset = application.volume.dataset
+                    name = dataset.metadata[u"name"]
+                    current_datasets_by_name[name] = dataset
+        for dataset in datasets_with_no_id:
+            matching = current_datasets_by_name.get(dataset.metadata[u"name"])
```

Comment:
I guess we should at least make a note here that this heuristic leaves something to be desired in the case where multiple volumes have been given the same name.  Long term this doesn't concern me much, since long term I think we're going to throw this heuristic away.  But I'm not quite sure how long it will be until we can do that.